File: bindings/ruby/test/conformance_test.rb
require 'minitest/autorun'
require 'json'
require_relative '../lib/strling/nodes'
require_relative '../lib/strling/ir'
require_relative '../lib/strling/core/hint_engine'

class ConformanceTest < Minitest::Test
  SPEC_DIR = File.expand_path('../../../../tests/spec', __FILE__)

  Dir.glob(File.join(SPEC_DIR, '*.json')).each do |file|
    base_name = File.basename(file, '.json')

    # Generate test name with special handling for semantic tests
    test_name = case base_name
                when 'semantic_duplicates'
                  'test_semantic_duplicate_capture_group'
                when 'semantic_ranges'
                  'test_semantic_ranges'
                else
                  "test_conformance_#{base_name.gsub(/[^a-zA-Z0-9_]/, '_')}"
                end
    
    # Pre-check to determine test type
    begin
      pre_spec = JSON.parse(File.read(file))
    rescue JSON::ParserError
      next
    end
    
    # Define test method based on spec type
    if pre_spec['input_ast'] && pre_spec['expected_ir']
      # Standard conformance test - capture pre_spec in closure to avoid re-reading file
      captured_spec = pre_spec
      define_method(test_name) do
        spec = captured_spec
        
        # Hydrate AST
        ast = Strling::Nodes::NodeFactory.from_json(spec['input_ast'])
        
        # Compile to IR
        ir = Strling::IR::Compiler.compile(ast)

        refute_nil ir, "Compilation returned nil"
        
        # Compare
        expected = spec['expected_ir']
        actual = serialize(ir)
        
        assert_equal expected, actual, "Mismatch in #{File.basename(file)}"
      end
    elsif pre_spec['expected_error']
      # Error test case
      captured_spec = pre_spec
      define_method(test_name) do
        spec = captured_spec
        if spec['input_ast']
            # If we have input_ast, try to compile and expect error
            ast = Strling::Nodes::NodeFactory.from_json(spec['input_ast'])
            assert_raises(StandardError) do
                Strling::IR::Compiler.compile(ast)
            end
        elsif spec['input_dsl'] && !spec['input_dsl'].empty?
            # Parser error test: parse input_dsl and verify error + hint
            begin
              require_relative '../lib/strling/core/parser'
              err = assert_raises(StandardError) do
                Strling::Core::Parser.new.parse(spec['input_dsl'])
              end
              assert_includes err.message, spec['expected_error'],
                "Error message mismatch.\n  Expected substring: #{spec['expected_error']}\n  Actual: #{err.message}"
              if spec['expected_hint'] && !spec['expected_hint'].empty? && err.respond_to?(:hint)
                assert_equal spec['expected_hint'], err.hint,
                  "Hint mismatch.\n  Expected: #{spec['expected_hint']}\n  Actual: #{err.hint}"
              end
            rescue LoadError
              # Parser not available, pass
              pass
            end
        else
            # Parser test (no AST), out of scope. Pass.
            pass
        end
      end
    end
  end

  def serialize(obj)
    case obj
    when Data
      obj.to_h.transform_keys(&:to_s).transform_values { |v| serialize(v) }.compact
    when Array
      obj.map { |v| serialize(v) }
    when Hash
      obj.transform_keys(&:to_s).transform_values { |v| serialize(v) }.compact
    else
      obj
    end
  end
end

File: bindings/dart/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:strling/strling.dart';
import 'package:strling/src/core/parser.dart' as parser_lib;
import 'package:test/test.dart';

void main() {
  // Adjust path logic to be robust
  // When running `dart test`, CWD is usually the package root (bindings/dart)
  var specPath = p.join(Directory.current.path, '../../tests/spec');
  var specDir = Directory(specPath);

  if (!specDir.existsSync()) {
    // Try relative to script if CWD is different
    // This might happen if running from workspace root
    specPath = p.join(Directory.current.path, 'tests/spec');
    specDir = Directory(specPath);
  }

  if (!specDir.existsSync()) {
    print(
        'Warning: Spec directory not found at ${specDir.path} or ../../tests/spec');
    return;
  }

  final files = specDir
      .listSync()
      .whereType<File>()
      .where((f) => f.path.endsWith('.json'));

  for (final file in files) {
    final filename = p.basename(file.path);
    final content = file.readAsStringSync();
    final json = jsonDecode(content) as Map<String, dynamic>;

    if (json.containsKey('expected_error')) {
      if (json.containsKey('input_ast') && !json.containsKey('input_dsl')) {
        test('Conformance (Error): $filename', () {
          print('=== RUN $filename');
          final inputAst = json['input_ast'] as Map<String, dynamic>;
          try {
            final node = Node.fromJson(inputAst);
            node.toIR();
            fail('Expected error but compilation succeeded');
          } catch (e) {
            print('    --- PASS: Caught expected error');
          }
        });
      } else if (json.containsKey('input_dsl') &&
          json.containsKey('expected_hint')) {
        test('Conformance (Parser Error): $filename', () {
          print('=== RUN $filename');
          final inputDsl = json['input_dsl'] as String;
          final expectedError = json['expected_error'] as String;
          final expectedHint = json['expected_hint'] as String;
          try {
            parser_lib.parse(inputDsl);
            fail('Expected error \'$expectedError\' but parsing succeeded');
          } on parser_lib.STRlingParseError catch (e) {
            expect(e.message, contains(expectedError),
                reason: 'Error message mismatch in $filename');
            expect(e.hint, equals(expectedHint),
                reason: 'Hint mismatch in $filename');
          }
        });
      } else {
        print('=== RUN $filename');
        print('    --- PASS: Parser test (no AST), out of scope');
      }
      continue;
    }

    if (!json.containsKey('input_ast') || !json.containsKey('expected_ir')) {
      continue;
    }

    test('Conformance: $filename', () {
      print('=== RUN $filename');
      final inputAst = json['input_ast'] as Map<String, dynamic>;
      final expectedIr = json['expected_ir'] as Map<String, dynamic>;

      try {
        final node = Node.fromJson(inputAst);
        final actualIr = node.toIR();
        expect(actualIr, equals(expectedIr));
      } catch (e, s) {
        print('Failed to process $filename');
        print(e);
        print(s);
        print('    --- FAIL: Exception thrown');
      }
    });
  }
}
